### PR TITLE
Emit iterative WASM for array_map (#480 PR 1)

### PR DESCRIPTION
**Pull request by aallan**
## Summary

- Replace the recursive prelude `array_map` / `array_map_go` with a single WAT `loop` driven by `call_indirect` on the closure. Shadow-stack usage is now O(1) regardless of input length — the old recursive version hit the 16K ceiling (post-#464) around 4K elements.
- Preserves the generic `forall<A, B>` signature through the existing `FunctionInfo` built-in registration; source programs are unchanged.
- Covered by 4 new tests in `TestArrayOperations`: 8K-element stress, Int→Bool type change, closure capturing outer variable, Int→String pair output.

## Discovered along the way

Filed #484 while debugging the stress test: the GC's object header size field is 16-bit, so allocations ≥65536 bytes make the sweeper link middle-of-payload bytes into the free list and clobber the live object. Test is capped at 8K Int elements (64,000 bytes) until #484 lands; full reproducer and fix options are in the issue.

Part of [#480](https://github.com/aallan/vera/issues/480) — PR 2 (`array_filter`) and PR 3 (`array_fold`) to follow.

## Test plan
- [x] `mypy vera/` clean (57 files)
- [x] `pytest tests/ -q` — 3,312 passed, 11 skipped
- [x] 19 `TestArrayOperations` tests pass including the 4 new ones
- [x] Pre-commit hooks (doc counts, CHANGELOG, allowlists, etc.) green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Array mapping now uses an iterative implementation, reducing shadow‑stack use to O(1) and improving large‑array reliability while preserving generic behaviour.

* **Known Issues**
  * Documented GC header limitation: 16‑bit size field can corrupt allocations above 65,535 bytes.

* **Tests**
  * Added five regression tests covering large arrays, mixed‑type maps, closure captures, pair‑typed results and zero‑length boundary cases.

* **Documentation**
  * Updated changelog, roadmap and testing metrics.

### CODE REVIEW

File: vera/wasm/calls_arrays.py [hunk 1]
```diff
@@ -574,3 +577,207 @@ def _translate_array_slice(
...
+        Needed so array_map knows the size / load / store ops for the
+        output element type.  Handles the common case of an anonymous
+        function literal (``fn(...) -> T { ... }``).  Returns ``None``
+        when the return type can't be inferred from the AST alone.
+        """
+        if isinstance(closure_arg, ast.AnonFn):
+            ret = closure_arg.return_type
+            if isinstance(ret, ast.NamedType):
+                return self._resolve_type_name_to_wasm_canonical(ret.name)
+        return None
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_arrays.py:595]: @coderabbitai Verified — not a real issue for the current grammar. I tried `array_map(xs, double)` where `double` is a named function, and the parser rejects it with E005: passing anything other than an `fn(...) -> T { ... }` literal as the closure argument isn't syntactically valid today. Every `array_map`/`array_fold` call in `examples/` and `tests/conformance/` uses an `AnonFn` literal, which is why the AnonFn-only branch is sufficient.

When first-class function references or bare-name higher-order calls land (future Stage 11+ work), that feature will need to teach this helper to consult the typechecker's resolved types. Right now there's no call shape for the fallback to fire on, so adding it would be dead code with no test to exercise it. Leaving as-is; happy to revisit when that grammar work lands.

**Comment by aallan** [comment]
@coderabbitai Verified — the coupling you describe is pre-existing and doesn't break any current test or example.

Traced `vera/prelude.py:590-593`:

```python
if not array_fn_names.issubset(user_names):          # triggers ~always
    if not array_alias_names.issubset(user_names):   # triggers unless user shadows aliases
        source_parts.append(_ARRAY_TYPE_ALIASES)     # injects all 3 aliases
    source_parts.append(_ARRAY_COMBINATORS)
```

To skip alias injection, a user program has to shadow every name in `array_fn_names`. Before this PR that meant shadowing all 5 (`array_map` + `array_filter` + `array_filter_go` + `array_fold` + `array_fold_go`); after the PR it means shadowing all 4 of the remaining recursive helpers. Either way it's a corner case: a user would have to both shadow `array_filter` and `array_fold` **and** reference `ArrayMapFn` in a type annotation to hit it.

Crucially, `array_map` itself no longer depends on `ArrayMapFn` — its full `forall<A, B>` signature lives in the `FunctionInfo` built-in registration in `environment.py`, so the built-in works whether or not the alias is injected. The alias is purely a user-facing convenience for spelling function types.

No current test, example, or conformance program hits that intersection. Decoupling would add a `_ARRAY_ALIASES_ONLY` emission branch with no test to exercise it, which is exactly the "dead code, no test" pattern we've been trimming elsewhere. Leaving as-is; happy to revisit when a real user program trips it, or as part of a future cleanup that refactors all three alias-bundle sites (Option, Result, Array) consistently.

File: tests/test_codegen_monomorphize.py [hunk 2]
```diff
@@ -895,3 +895,99 @@ def test_array_fold_type_check(self) -> None:
...
 """
         _compile_ok(source)
+
+    # ----------------------------------------------------------------
+    # array_map — regression tests for the iterative implementation
+    # (#480).  These exercise paths the existing tests above don't:
+    # large inputs (stress the loop, not the recursion), closures
+    # that capture outer variables, A != B with pair output, and
+    # scalar Int → scalar Bool type change.
+    # ----------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_monomorphize.py:905]: @coderabbitai Fixed in 761658f — added `test_array_map_empty_input` to the regression block, verifying that `array_range(0, 0)` → `array_map(...)` → `array_length(...)` returns 0.  Exercises the loop's first-iteration `idx >= arr_len` guard, confirms the closure is never invoked, and confirms the output array's length is 0.

File: tests/test_codegen_monomorphize.py [hunk 3]
```diff
@@ -895,3 +895,99 @@ def test_array_fold_type_check(self) -> None:
...
+        one converts to Bool, which has a different WASM type (i32) and
+        different element width (1 byte vs 8) from Int.  The store ops
+        must pick up the B-sized layout, not reuse A's.
+        """
+        source = """\
+public fn main(-> @Int)
+  requires(true) ensures(true) effects(pure)
+{
+  let @Array<Bool> = array_map([0, 1, 2, 3], fn(@Int -> @Bool) effects(pure) { @Int.0 > 1 });
+  -- [false, false, true, true] → sum of true-indices = 2 + 3 = 5
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_codegen_monomorphize.py]: @coderabbitai Fixed in 761658f — comment now reads `-- [false, false, true, true] → sum of 1-for-each-true = 0+0+1+1 = 2`, which matches the `assert _run(source, fn="main") == 2` on the following line.

File: tests/test_prelude.py [hunk 4]
```diff
@@ -43,8 +43,11 @@ def _alias_names(prog: ast.Program) -> set[str]:
...
 _RESULT_FN_NAMES = {"result_unwrap_or", "result_map"}
+# ``array_map`` is emitted as iterative WASM by codegen (#480);
+# it no longer has a prelude-injected recursive implementation.
+# ``array_filter`` / ``array_fold`` remain recursive prelude functions
+# until their own iterative migration lands.
 _ARRAY_FN_NAMES = {
-    "array_map", "array_map_go",
     "array_filter", "array_filter_go",
     "array_fold", "array_fold_go",
 }
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_prelude.py:53]: @coderabbitai Fixed in 761658f — `test_array_operations_injected` now explicitly asserts `"array_map" not in names` and `"array_map_go" not in names` alongside the existing `_ARRAY_FN_NAMES.issubset(names)` check, with a docstring explaining it's a guard against accidental re-injection if `_ARRAY_FN_NAMES` is ever widened without updating the prelude.

File: vera/prelude.py [hunk 5]
```diff
@@ -586,8 +560,11 @@ def inject_prelude(program: ast.Program) -> None:
     option_alias_names = {"OptionMapFn", "OptionBindFn"}
     result_fn_names = {"result_unwrap_or", "result_map"}
     result_alias_names = {"ResultMapFn"}
+    # array_map is a built-in emitted as iterative WASM (#480); it has
+    # no prelude body any more.  array_filter / array_fold are still
+    # injected as recursive helpers until their own iterative migration
+    # lands.
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/prelude.py:568]: @coderabbitai Fixed in 761658f — `inject_prelude` docstring now reads: `Array operations (array_filter, array_fold) — always injected as recursive helpers until their own iterative migration lands. array_map is emitted as iterative WASM by codegen (#480) and has no prelude body.`

File: vera/wasm/calls_arrays.py [hunk 6]
```diff
@@ -574,3 +577,207 @@ def _translate_array_slice(
...
+            ;;   push fn (env); load arr[idx]; push fn.func_table_idx
+            ;;   call_indirect (type $closure_sig_N)
+            ;;   save result; compute dst_slot; store
+            ;; push (dst, len)
+
+        Note: the GC's header size field is 16-bit, so allocations
+        larger than 65535 bytes currently trigger a separate (pre-
+        existing) sweeper bug.  Callers should keep output arrays
+        under ~8K elements for Int/Nat/Float64 or under ~16K for
+        Bool/Byte until that limit is raised.
```
[the remark this answers is not in the document]
Comment (aallan) [reply vera/wasm/calls_arrays.py:626]: @coderabbitai Declined — a per-site trap in `array_map` wouldn't actually make the system safer.  The same silent-corruption risk applies to `array_range`, `array_concat`, `array_append`, and `array_slice` today (any of them can allocate >65535 bytes), so adding the guard here and nowhere else would just shift the landmine without defusing it.

The correct fix is #484 (widen the 16-bit GC-header size field to 31 bits in `vera/codegen/assembly.py` — a ~5-line change across two `i32.const 65535 / i32.and` masks in `$alloc` and three in `$gc_collect`). That cures every allocation site in one place and lets this test go back to 10K+ elements. I'd rather land that as the immediate follow-up than sprinkle bandaids across combinators.

**Comment by aallan** [comment]
@coderabbitai Fixed in 839e41d — the closure-capture test comment now matches the assertion.  `[1, 2, 3][2] = 3`, so `closure returns element + captured = 3 + 100 = 103`, which is exactly what the `assert _run(source, fn="main") == 103` checks. Your suggested diff had the arithmetic right (element 3, not element 2); I expanded it slightly to spell out the array-index step for future readers:

```python
# Outer @Int.0 = 100 (captured); [1, 2, 3][2] = 3;
# closure returns element + captured = 3 + 100 = 103.
assert _run(source, fn="main") == 103
```
